File: structural design patterns/Adapter Pattern/C++/adapter_pattern.cpp
/*
In this example, we have a Target Interface called "Shape" that defines a method "draw()".
We also have an Adaptee class called "LegacyRectangle" that has a different interface 
with a method "drawRectangle()". The Adapter class "RectangleAdapter" implements the 
Target Interface and adapts the Adaptee's interface to match it.

When the client calls the "draw()" method on the Adapter, it internally calls the 
"drawRectangle()" method of the Adaptee, allowing the two incompatible interfaces to 
work together seamlessly.
*/

#include <bits/stdc++.h>
using namespace std;

// Target Interface
class Shape {
public:
    virtual void draw() = 0;
};

// Concrete implementation of the Target Interface
class Circle : public Shape {
public:
    void draw() override {
        cout << "Drawing a circle." << endl;
    }
};

// Adaptee or Existing class with an incompatible interface
class LegacyRectangle {
public:
    void drawRectangle() {
        cout << "Drawing a rectangle using LegacyRectangle." << endl;
    }
};

// Adapter or Wrapper class that implements the Target Interface and adapts the Adaptee's interface
class RectangleAdapter : public Shape {
private:
    /*
    The RectangleAdapter class has a private member variable that holds a pointer 
    to the LegacyRectangle object.
    */
    LegacyRectangle* legacyRectangle;
public:
    RectangleAdapter() {
        // Initialize the legacy rectangle object in the constructor
        this->legacyRectangle = new LegacyRectangle();
    }
    
    void draw() override {
        this->legacyRectangle->drawRectangle();
    }
};

// Client code
// Draw Service class that uses the Target Interface
class DrawService {
    Shape* shape;
public: 
    // Constructor to set the shape
    DrawService(Shape* shape){
        this->shape = shape;
    }

    // Method to draw the shape
    void drawShape() {
        this->shape->draw();
    }
};

int main() {
    // Create a Circle object and use it with the DrawService
    Circle circle;
    DrawService drawCircle(&circle);
    drawCircle.drawShape();

    // Create a RectangleAdapter object and use it with the DrawService
    RectangleAdapter rectangleAdapter;
    DrawService drawRectangle(&rectangleAdapter);
    drawRectangle.drawShape();

    return 0;
}

File: structural design patterns/Adapter Pattern/Java/adapter_pattern.java
/*
In this example, we have a Target Interface called "Shape" that defines a method "draw()".
We also have an Adaptee class called "LegacyRectangle" that has a different interface 
with a method "drawRectangle()". The Adapter class "RectangleAdapter" implements the 
Target Interface and adapts the Adaptee's interface to match it.

When the client calls the "draw()" method on the Adapter, it internally calls the 
"drawRectangle()" method of the Adaptee, allowing the two incompatible interfaces to 
work together seamlessly.
*/

import java.util.*;

// Target Interface
interface Shape{
    public void draw();
}

// Concrete implementation of the target interface
class Circle implements Shape{
    @Override
    public void draw(){
        System.out.println("Drawing a circle");
    }
}

// Adaptee or Existing class with an incompatible interface
class LegacyRectangle{
    public void drawRectangle(){
        System.out.println("Drawing a rectangle using LegacyRectangle");
    }
}

// Adapter or Wrapper class that implements the target interface and adapts adaptee's interafce
class RectangleAdapter implements Shape{
    /*
    The RectangleAdapter class has a private member variable that holds a pointer 
    to the LegacyRectangle object.
    */
    private LegacyRectangle legacyRectangle = new LegacyRectangle();

    @Override
    public void draw(){
        this.legacyRectangle.drawRectangle();
    }
}

// Client code
// Draw service class that uses the target interface
class DrawService{
    private Shape shape;

    // Constructor to set the shape
    DrawService(Shape shape){
        this.shape=shape;
    }

    // Method to draw the shape
    public void drawShape(){
        this.shape.draw();
    }
}

public class Main{
    public static void main(String[] args){
        // Create a circle object and use it with the DrawService
        Circle circle = new Circle();
        DrawService drawCircle = new DrawService(circle);
        drawCircle.drawShape();

        // Create a RectancleAdapter object and use it with the DrawService
        RectangleAdapter rectangleAdapter = new RectangleAdapter();
        DrawService drawRectangle = new DrawService(rectangleAdapter);
        drawRectangle.drawShape();
    }
}